### DISCUSSION TITLE
Subnetworking and Step Multiplication

### DISCUSSION THREAD

**Discussion by CAP1Sup** [CAP1Sup/Intellistep#9] [Ideas]
Have any ideas? Post them here!

**Comment by lav8org** [comment]
I'm wondering if a standalone closed loop increases "where" accuracy at the expense of "when" accuracy? It seems to me that any slop in the PID loop is going to mess up axis synchronization, producing rounded corners and arced lines. If so, assuming the PID can't be sufficiently tuned, presumably the only fix is to get the controllers talking to each other? To that end I can roughly see two possibilities: 
1. Subnetwork: every S42B coordinates with each other over CAN, perhaps with step/dir signals received by each forwarded to one that manages them. 
Pro: still standalone, motherboard agnostic, broadest application potential. 
Con: seems inordinately complex. I'm surprisingly still quite naive about the details of step/dir, but I'd guess maintaining the original intent/sync of the signals may not be possible this way. Also further decoupling of motor state from motherboard state--lots more failure modes, and maybe requires all constantly changing axes (x y e) to be looped
2. Network: each is managed by the motherboard, whether just providing feedback or skipping the step/dir interface entirely and moving all logic to the CAN bus. 
Pro: unified machine state, less faffing about remanaging already carefully coordinated signals.
Con: motherboard hardware support--I haven't seen any with a dedicated CAN port, is it possible to retask some pins? Maybe it would need to be serial? Either way there's the issue of configuration breadth. Also the potential complexity of the involvement of other project/s.

As it happens Marlin already has code for closed loop meant for a specific linear magnetic encoder setup; AFAICT the associated hardware never caught on and is no longer obtainable (the originator's store as linked in the Marlin source still has a page for it, but no stock). That code might be a big foothold.

**Reply by CAP1Sup** [reply]
I was hoping to move the whole closed loop idea away from the way that BTT did it, which is wrong in my opinion. Closed loop should only correct if the motor is out of it's set place. That's it. BTT tries to run a whole PID loop with complex tuning all of the time, but there doesn't seem to be a reason for it. The Intellistep project only attempts to correct the motor when required. Otherwise, it should blend right in with other stepper drivers in terms of performance.

As for your whole idea of networking, that's definitely possible. However, only settings should be sent over the CAN bus as the stepper timing would be a nightmare. It's something that we could definitely try, but for now I think that I'm going to stray away from stepping over Serial/CAN. As for the whole idea of a subnet, absolutely. I'm going to add a serial command that will send out a CAN bus message soon. This would allow one controller or computer to talk to all of the modules. However, I don't have a USB to CAN converter, so I'm going to have to do testing with multiple modules. For now, I just want to get the basic features down like stepping so that this project can get off the ground while people are still interested. No sense in adding shiny paint to a car if it's engine doesn't work. I'm really just struggling with the encoder yet, once that is complete I should be able to do the final testing to confirm the stepping coils are correctly managed.

**Comment by maracMB** [comment]
Hello, this is a more "formal" request to implement step factoring/multiplication that @CAP1Sup already did.

Reason is, that with fdm 3d printers, with multiple individual heads, there can be some rationalization made with extruder drives in particular to save hundreds of money units on expensive mainboards (duet2 vs duet3 for example) and to also rationalize computational resources.
So with extruders on multihead machines is; Either all run, either they run in pairs, none runs... but they run the same none the less. So, there is an option to run feeds to drivers of these motors through demultiplexer and that enables cheap expansion with external stepper drivers.

Problem arrises with the use of different extruders (geared, geared differently, non-geared, etc) - it would not work unless there is some kind of step factoring done for each motor.

Having a separate control board that recalculates steps either way, the factored steps are possible and thus enabling multiple motors to run at multiple steps/mm from one stepper driver port on mainboard side.

**Comment by CAP1Sup** [comment]
> Hello, this is a more "formal" request to implement step factoring/multiplication that @CAP1Sup already did.
> 
> Reason is, that with fdm 3d printers, with multiple individual heads, there can be some rationalization made with extruder drives in particular to save hundreds of money units on expensive mainboards (duet2 vs duet3 for example) and to also rationalize computational resources.
> So with extruders on multihead machines is; Either all run, either they run in pairs, none runs... but they run the same none the less. So, there is an option to run feeds to drivers of these motors through demultiplexer and that enables cheap expansion with external stepper drivers.
> 
> Problem arrises with the use of different extruders (geared, geared differently, non-geared, etc) - it would not work unless there is some kind of step factoring done for each motor.
> 
> Having a separate control board that recalculates steps either way, the factored steps are possible and thus enabling multiple motors to run at multiple steps/mm from one stepper driver port on mainboard side.

So it's all implemented. There's two factors to consider: understand how the values translate and a possible full step limit. The values work in relation to the normal base microstepping. For example, if you have a 1:1 stepper and a geared down 1:2 stepper, you would want to set the multiplier on the 1:2 to 2. I don't know what will happen if the value exceeds full steps (I'm going to work to make sure that's not an issue and it would just picked up by the normal routine). Also, the motor will not do half microsteps. It'll just round the value off. You should base your multipliers on the stepper that has to move the least. In the previous example, setting the 1:1 stepper at .5 instead would result in unpredictable behavior. Just be warned when you are using it in the future
